plugins/externalwrench/src/ExternalWrench.cc

```diff
@@ -0,0 +1,181 @@
+#include "ExternalWrench.hh"
+
+int ExternalWrench::count = 0;
+
+//Initializing wrench command
+ExternalWrench::ExternalWrench()
+{
+    model_has_link = false;
+
+    //srand(boost::lexical_cast<float>(std::time(NULL)));
+    srand(rand()%100);
+    color[0] = static_cast <float> (rand()) / static_cast <float> (RAND_MAX);
+    color[1] = static_cast <float> (rand()) / static_cast <float> (RAND_MAX);
+    color[2] = static_cast <float> (rand()) / static_cast <float> (RAND_MAX);
+    color[3] = static_cast <float> (rand()) / static_cast <float> (RAND_MAX);
+
+    wrenchPtr->link_name = "world";
+    wrenchPtr->force.resize(3,0);
+    wrenchPtr->torque.resize(3,0);
+    wrenchPtr->duration = 0.0;
+
+    count++;
+    tick = yarp::os::Time::now();
+    duration_done = false;
+}
+
+bool ExternalWrench::getLink()
+{
+    // Get the exact link with only link name instead of full_scoped_link_name
+    links = this->model->GetLinks();
+    for(int i=0; i < links.size(); i++)
+    {
+        std::string candidate_link_name = links[i]->GetScopedName();
+
+        // hasEnding compare ending of the condidate model link name to the given link name, in order to be able to use unscoped names
+        if(GazeboYarpPlugins::hasEnding(candidate_link_name, wrenchPtr->link_name))
+        {
+            link = links[i];
+            break;
+        }
+    }
+
+    if(!link)
+     {
+        yError()  << "externalWrench: " << wrenchPtr->link_name << " link is not found";
+        return false;
+    }
+
+    model_has_link = true;
+    return true;
+}
+
+void ExternalWrench::setVisual()
+{
+    //Wrench Visual
+    m_node = transport::NodePtr(new gazebo::transport::Node());
+    this->m_node->Init(model->GetWorld()->Name());
+    m_visPub = this->m_node->Advertise<msgs::Visual>("~/visual",100);
+
+    // Set the visual's name. This should be unique.
+    std::string visual_name = "__" + wrenchPtr->link_name + "__CYLINDER_VISUAL__" + boost::lexical_cast<std::string>(count);
+    m_visualMsg.set_name (visual_name);
+
+    // Set the visual's parent. This visual will be attached to the parent
+    m_visualMsg.set_parent_name(model->GetScopedName());
+
+    // Create a cylinder
+    msgs::Geometry *geomMsg = m_visualMsg.mutable_geometry();
+    geomMsg->set_type(msgs::Geometry::CYLINDER);
+    geomMsg->mutable_cylinder()->set_radius(0.0075);
+    geomMsg->mutable_cylinder()->set_length(.30);
+
+    // Don't cast shadows
+    m_visualMsg.set_cast_shadows(false);
+}
+
+bool ExternalWrench::setWrench(physics::ModelPtr& _model,yarp::os::Bottle& cmd)
+{
+    model = _model;
+
+    //get link name from command
+    std::string linkName = cmd.get(0).asString();
+    if (linkName.find_last_of("::") != 0) {
+        wrenchPtr->link_name = linkName.substr(linkName.find_last_of("::")+1); //case of scoped link name
```

Comment:
Probably this is not needed, we can simply use `wrenchPtr->link_name = linkName` also in case of scoped `linkName`.

If a scoped link name is passed we don't want to cancel submodel names which can be useful when multiple models are used (e.g. they exists both `iCub::iCub_1::l_hand` and `iCub::iCub_2::l_hand`). [hasEnding](https://github.com/robotology/gazebo-yarp-plugins/pull/418/files#diff-bdf55637a7d0134a9ccb12f75775a362R36) should handle also scoped names.

**e.g.**
To make a practical example, if we have to models with `l_hand`, we could have two links with names `iCub::iCub_1::l_hand` and `iCub::iCub_2::l_hand`.
If now we want to target the hand of `iCub::iCub_2`, we should be able to use the scoped name `iCub_2::l_hand`. 
But the way it is implemented now it will just search for `l_hand` ending with `hasEnding` which could result in the other model link `iCub::iCub_2::l_hand`. 
If instead we simply use the scoped name as `linkname`, when `hasEnding` will be called, the correct link `iCub::iCub_2::l_hand` would be found.

Reply:
> But the way it is implemented now it will just search for l_hand ending with hasEnding which could result in the other model link iCub::iCub_2::l_hand.

I believe this will not happen because this plugin is a model plugin. So, it will only consider the model to which it is attached.

> Probably this is not needed, we can simply use wrenchPtr->link_name = linkName also in case of scoped linkName.

You are correct. I put this check in place to ensure back compatibility for current users who use command format link as `iCub::l_hand`. But the `hasEnding` function will take care of that.

Reply:
Done